pointcloud/basic_octree.go

```diff
@@ -390,6 +390,90 @@ func (octree *BasicOctree) FinalizeAfterReading() (PointCloud, error) {
 	return octree, nil
 }
 
+// PointsCollidingWith returns all points in the octree that collide with any of the given geometries.
+// A point is considered colliding if it meets the confidence threshold and is within the collision buffer distance.
+func (octree *BasicOctree) PointsCollidingWith(geometries []spatialmath.Geometry, collisionBufferMM float64) []r3.Vector {
+	// Early exit if this octree region has no points above confidence threshold
+	if octree.MaxVal() < octree.confidenceThreshold {
+		return nil
+	}
+
+	var collidingPoints []r3.Vector
+
+	switch octree.node.nodeType {
+	case internalNode:
+		// Create a bounding box for this octree region
+		ocbox, err := spatialmath.NewBox(
+			spatialmath.NewPoseFromPoint(octree.center),
+			r3.Vector{
+				X: octree.sideLength + collisionBufferMM,
+				Y: octree.sideLength + collisionBufferMM,
+				Z: octree.sideLength + collisionBufferMM,
+			},
+			"",
+		)
+		if err != nil {
+			return nil
+		}
+
+		// Check if any geometry intersects with this octree region
+		intersects := false
+		for _, geom := range geometries {
+			collides, err := geom.CollidesWith(ocbox, collisionBufferMM)
+			if err == nil && collides {
+				intersects = true
+				break
+			}
+		}
+
+		// If no geometry intersects this region, skip all children
+		if !intersects {
+			return nil
+		}
+
+		// Recursively check children and collect results
+		for _, child := range octree.node.children {
+			childPoints := child.PointsCollidingWith(geometries, collisionBufferMM)
+			collidingPoints = append(collidingPoints, childPoints...)
+		}
+
+	case leafNodeEmpty:
+		// Empty leaf has no points
+		return nil
+
+	case leafNodeFilled:
+		// Check confidence threshold
+		if octree.node.point.D.HasValue() && octree.node.point.D.Value() < octree.confidenceThreshold {
+			return nil
+		}
+
+		// Create a point geometry for collision checking
+		pointGeom := spatialmath.NewPoint(octree.node.point.P, "")
+
+		// Check collision with each geometry
+		for _, geom := range geometries {
+			collides, err := pointGeom.CollidesWith(geom, collisionBufferMM)
+			if err == nil && collides {
+				collidingPoints = append(collidingPoints, octree.node.point.P)
+				break // Point collides with at least one geometry, no need to check others
```

Comment:
will this not skip some of the collision points from other geometries? Or is this function not intended to return _all_ of the collision points?

Reply:
ah I see